lib/api/apiUtils/object/createAndStoreObject.js

```diff
@@ -0,0 +1,158 @@
+        });
+    }
+    const backendInfo = backendInfoObj.backendInfo;
+
+    const requestLogger =
+        logger.newRequestLoggerFromSerializedUids(log.getSerializedUids());
+    return async.waterfall([
+        function storeData(next) {
+            if (size === 0) {
+                metadataStoreParams.contentMD5 = constants.emptyFileMd5;
+                return next(null, null, null);
+            }
+            return dataStore(objectKeyContext, cipherBundle, request, size,
+                    streamingV4Params, backendInfo, log, next);
+        },
+        function processDataResult(dataGetInfo, calculatedHash, next) {
+            if (dataGetInfo === null || dataGetInfo === undefined) {
+                return next(null, null);
+            }
+            // So that data retrieval information for MPU's and
+            // regular puts are stored in the same data structure,
+            // place the retrieval info here into a single element array
+            const { key, dataStoreName } = dataGetInfo;
+            const dataGetInfoArr = [{ key, size, start: 0, dataStoreName }];
+            if (cipherBundle) {
+                dataGetInfoArr[0].cryptoScheme = cipherBundle.cryptoScheme;
+                dataGetInfoArr[0].cipheredDataKey =
+                    cipherBundle.cipheredDataKey;
+            }
+            metadataStoreParams.contentMD5 = calculatedHash;
+            return next(null, dataGetInfoArr);
+        },
+        function getVersioningInfo(infoArr, next) {
+            return versioningPreprocessing(bucketName, bucketMD,
+                metadataStoreParams.objectKey, objMD, reqVersionId, log,
+                (err, options) => next(err, options, infoArr));
+        },
+        function createNullVerDeleteArray(options, infoArr, next) {
+            if (!options.deleteNullVersionData) {
+                return next(null, options, infoArr);
+            }
+            // When options.deleteNullVersionData is true, need to get
+            // location info of null version for deletion. Only applies when
+            // there is pre-existing null version that is not the latest version
+            // and versioning is suspended.
+            const params = { versionId: options.nullVersionId };
+            return metadata.getObjectMD(bucketName, objectKey,
+                params, log, (err, nullObjMD) => {
+                    if (err) {
+                        log.debug('err from metadata', {
+                            error: err,
+                        });
+                        return next(err);
+                    }
+                    if (nullObjMD.location) {
+                        const nullDataToDelete =
+                            Array.isArray(nullObjMD.location) ?
+                            nullObjMD.location : [nullObjMD.location];
+                        dataToDelete = dataToDelete ?
+                            dataToDelete.push(nullDataToDelete) :
```

Comment:
I need to update this - nullDataToDelete (when overwriting a null version that is not the latest version) should not happen when there is dataToDelete (overwriting data from the latest version).